src/catalog/catalog.cpp

```diff
@@ -930,6 +935,287 @@ std::shared_ptr<SystemCatalogs> Catalog::GetSystemCatalogs(
+      // Step 1: build empty table with new schema
+      bool own_schema = true;
+      bool adapt_table = false;
+      auto new_table = storage::TableFactory::GetDataTable(
+          database_oid, table_oid,
+          catalog::Schema::CopySchema(new_schema.get()), old_table->GetName(),
+          DEFAULT_TUPLES_PER_TILEGROUP, own_schema, adapt_table);
+      // Step 2: Copy indexes
+      auto old_index_oids = pg_index->GetIndexObjects(table_oid, txn);
+      for (auto index_oid_pair : old_index_oids) {
+        oid_t index_oid = index_oid_pair.first;
+        // delete record in pg_index
+        pg_index->DeleteIndex(index_oid, txn);
+        // Check if all indexed columns still exists
+        auto old_index = old_table->GetIndexWithOid(index_oid);
+        bool index_exist = true;
+        std::vector<oid_t> new_key_attrs;
+
+        for (oid_t column_id : old_index->GetMetadata()->GetKeyAttrs()) {
+          bool is_found = false;
+          std::string column_name = old_schema->GetColumn(column_id).GetName();
+          oid_t i = 0;
+          for (auto new_column : new_schema->GetColumns()) {
+            if (column_name == new_column.GetName()) {
+              is_found = true;
+              new_key_attrs.push_back(i);
+              break;
+            }
+            i++;
+          }
+          if (!is_found) {
+            index_exist = false;
+            break;
+          }
+        }
+        if (!index_exist) continue;
+
+        // construct index on new table
+        auto index_metadata = new index::IndexMetadata(
+            old_index->GetName(), index_oid, table_oid, database_oid,
+            old_index->GetMetadata()->GetIndexType(),
+            old_index->GetMetadata()->GetIndexConstraintType(),
+            new_schema.get(),
+            catalog::Schema::CopySchema(new_schema.get(), new_key_attrs),
+            new_key_attrs, old_index->GetMetadata()->HasUniqueKeys());
+
+        std::shared_ptr<index::Index> new_index(
+            index::IndexFactory::GetIndex(index_metadata));
+        new_table->AddIndex(new_index);
+
+        // reinsert record into pg_index
+        pg_index->InsertIndex(
+            index_oid, old_index->GetName(), table_oid, schema_name,
+            old_index->GetMetadata()->GetIndexType(),
+            old_index->GetMetadata()->GetIndexConstraintType(),
+            old_index->GetMetadata()->HasUniqueKeys(), new_key_attrs,
+            pool_.get(), txn);
+      }
+      std::unique_ptr<executor::ExecutorContext> context(
+          new executor::ExecutorContext(txn, {}));
+      // Step 3: build column mapping between old table and new table
+      // we're using column name as unique identifier
+      std::vector<oid_t> old_column_ids;
+      std::unordered_map<oid_t, oid_t> column_map;
+      for (oid_t old_column_id = 0;
+           old_column_id < old_schema->GetColumnCount(); old_column_id++) {
+        old_column_ids.push_back(old_column_id);
+        for (oid_t new_column_id = 0;
```

Comment:
[inefficient implementation]
It is not efficient to get a matched name in a for loop. It is better to use a hashmap to do such a thing.

Reply:
Sure, thanks for your suggestion.